mqtt/mqtt-broker/benches/file_persist_bench.rs

```diff
@@ -0,0 +1,181 @@
+use std::any::type_name;
+use std::collections::{HashMap, VecDeque};
+use std::iter::FromIterator;
+
+use bytes::Bytes;
+use criterion::*;
+use mqtt3::proto::{Publication, QoS};
+use mqtt_broker::{
+    BincodeFormat, BrokerState, ClientId, ConsolidatedStateFormat, FileFormat, FilePersistor,
+    Persist, SessionState,
+};
+use tempfile::TempDir;
+use tokio::runtime::Runtime;
+
+fn test_write<F>(
+    c: &mut Criterion,
+    num_clients: u32,
+    num_unique_messages: u32,
+    num_shared_messages: u32,
+    num_retained: u32,
+    format: F,
+) where
+    F: FileFormat + Clone + Send + 'static,
+{
+    let name = format!(
+        "{}: Write {} unique and {} shared messages for {} sessions with {} retained messages",
+        type_name::<F>(),
+        num_unique_messages,
+        num_shared_messages,
+        num_clients,
+        num_retained
+    );
+
+    c.bench_function(&name, |b| {
+        b.iter_batched(
+            || {
+                let state = make_fake_state(
+                    num_clients,
+                    num_unique_messages,
+                    num_shared_messages,
+                    num_retained,
+                );
+
+                let tmp_dir = TempDir::new().unwrap();
+                let path = tmp_dir.path().to_owned();
+                let persistor = FilePersistor::new(path, format.clone());
+                let rt = Runtime::new().unwrap();
+
+                (state, persistor, rt, tmp_dir)
+            },
+            |(state, mut persistor, mut rt, _tmp_dir)| {
+                rt.block_on(async { persistor.store(state).await })
+            },
+            BatchSize::SmallInput,
+        );
+    });
+}
+
+fn test_read<F>(
+    c: &mut Criterion,
+    num_clients: u32,
+    num_unique_messages: u32,
+    num_shared_messages: u32,
+    num_retained: u32,
+    format: F,
+) where
+    F: FileFormat + Clone + Send + 'static,
+{
+    let name = format!(
+        "{}: Read {} unique and {} shared messages for {} sessions with {} retained messages",
+        type_name::<F>(),
+        num_unique_messages,
+        num_shared_messages,
+        num_clients,
+        num_retained
+    );
+
+    c.bench_function(&name, |b| {
+        b.iter_batched(
+            || {
+                let state = make_fake_state(
+                    num_clients,
+                    num_unique_messages,
+                    num_shared_messages,
+                    num_retained,
+                );
+
+                let tmp_dir = TempDir::new().unwrap();
+                let path = tmp_dir.path().to_owned();
+                let mut persistor = FilePersistor::new(path, format.clone());
+
+                Runtime::new()
```

Comment:
is it possible to write perf test with some macro like `#[tokio:test]` ?

Reply:
I wasn't able to find a way. I think async/await is new enough

Reply:
Ok then. Will wait until everyone caught up